src/noConflictingLifecycleHooksRule.ts

```diff
@@ -0,0 +1,100 @@
+import { sprintf } from 'sprintf-js';
+import { IRuleMetadata, RuleFailure, RuleWalker, Utils } from 'tslint';
+import { AbstractRule } from 'tslint/lib/rules';
+import { ClassDeclaration, SourceFile } from 'typescript';
+import {
+  getDeclaredLifecycleInterfaces,
+  getDeclaredLifecycleMethods,
+  LifecycleInterfaceKeys,
+  LifecycleInterfaces,
+  LifecycleMethodKeys,
+  LifecycleMethods
+} from './util/utils';
+
+const LIFECYCLE_INTERFACE_CONFLICT_MAPPER: Partial<Record<LifecycleInterfaceKeys, LifecycleInterfaceKeys>> = {
+  [LifecycleInterfaces.DoCheck]: LifecycleInterfaces.OnChanges
+};
+const LIFECYCLE_METHOD_CONFLICT_MAPPER: Partial<Record<LifecycleMethods, LifecycleMethods>> = {
+  [LifecycleMethods.ngDoCheck]: LifecycleMethods.ngOnChanges
+};
+
+export const getFailureMessage = (
+  methodOne: LifecycleInterfaceKeys | LifecycleMethodKeys,
+  methodTwo: LifecycleInterfaceKeys | LifecycleMethodKeys,
+  className: string,
+  errorMessage: typeof Rule.FAILURE_STRING_INTERFACE_HOOK | typeof Rule.FAILURE_STRING_METHOD_HOOK
+): string => sprintf(errorMessage, methodOne, methodTwo, className);
+
+export class Rule extends AbstractRule {
+  static readonly metadata: IRuleMetadata = {
+    description: 'Ensures that directives not implement conflicting lifecycle hooks.',
+    descriptionDetails: 'See more at https://angular.io/api/core/DoCheck#description.',
+    options: null,
+    optionsDescription: 'Not configurable.',
+    rationale: Utils.dedent`
+      A directive typically should not use both DoCheck and OnChanges to respond
+      to changes on the same input, as ngOnChanges will continue to be called when the
+      default change detector detects changes.
+    `,
+    ruleName: 'no-conflicting-lifecycle-hooks',
+    type: 'functionality',
+    typescriptOnly: true
+  };
+
+  static readonly FAILURE_STRING_INTERFACE_HOOK = 'Implementing %s and %s hooks in class %s is not recommended';
+  static readonly FAILURE_STRING_METHOD_HOOK = 'Declaring %s and %s method hooks in class %s is not recommended';
+
+  apply(sourceFile: SourceFile): RuleFailure[] {
+    return this.applyWithWalker(new NoConflictingLifecycleInterfacesWalker(sourceFile, this.getOptions()));
+  }
+}
+
+export class NoConflictingLifecycleInterfacesWalker extends RuleWalker {
+  visitClassDeclaration(node: ClassDeclaration): void {
+    this.validateInterfaces(node);
+    this.validateMethods(node);
+    super.visitClassDeclaration(node);
+  }
+
+  private validateInterfaces(node: ClassDeclaration): void {
+    const declaredLifecycleInterfaces = getDeclaredLifecycleInterfaces(node);
+
+    for (const key of Object.keys(LIFECYCLE_INTERFACE_CONFLICT_MAPPER)) {
```

Comment:
Let's drop the for loop here, we don't really need it, nor `LIFECYCLE_INTERFACE_CONFLICT_MAPPER`

Reply:
Done.